# <font color=blue>Linear Regression Model Building Exercise </font>
### <font color=amber>Level. Intermediate</font>
This exercise consists of several phases designed to guide you through the full machine learning workflow.

**Phase 1: Understand the Problem**
- Define the business or prediction goal.
- Identify the target variable.
- Identify the input features.
- Determine whether this is a regression or classification problem.

**Phase 2: Explore the Data (Exploratory Data Analysis - EDA)**
- Load and inspect the dataset.
- Check data types.
- Handle missing values.
- Identify outliers.
- Generate summary statistics.
- Visualize relationships (e.g., scatter plots, correlation matrix).

**Phase 3: Prepare the Data**
- Select relevant features.
- Encode categorical variables (if any).

**Scale/normalize features if needed.**
- Split the dataset into training and testing sets (e.g., 80/20 split).

**Phase 4: Build a Linear Regression Model**
- Train a Linear Regression model on the training data.
- Examine model coefficients.
- Understand the intercept and feature importance.

**Phase 5: Evaluate the Linear Regression Model**
- Make predictions on the test set.
- Calculate evaluation metrics:
  - MAE (Mean Absolute Error)
  - RMSE (Root Mean Squared Error)
  - R² (R-squared)
-Compare training vs. testing performance (check for overfitting).

# <font color=blue> LinearRegression </font>
**👉 What it is:**
Basic model that finds a straight-line relationship between inputs (features) and output.

**✅ When to use:**
- You have **simple relationships**
- Features are **not highly correlated**
- You want **easy interpretability**

**🧠 Real-life examples:**
- Predicting **house price** based on size, bedrooms, location
- Estimating **salary** from years of experience
- Forecasting **sales** from advertising spend

**⚠️ Limitation:**
- Breaks when there are **too many features** or **multicollinearity (features highly correlated)**

### Generic libraries (e.g. Math, Plot etc.)

In [ ]:
# Import core libraries for numerical computing, data analysis, visualization, and suppress warnings

### ML model libraries (e.g. Linear regression, Classification etc.)

In [ ]:
# import model libraries

In [ ]:
# Load Diamonds dataset
# diamonds dataset: https://ggplot2.tidyverse.org/reference/diamonds.html


In [ ]:
# Checking the data

,carat,cut,color,clarity,depth,table,price,x,y,z
0,0.23,Ideal,E,SI2,61.5,55.0,326,3.95,3.98,2.43
1,0.21,Premium,E,SI1,59.8,61.0,326,3.89,3.84,2.31
2,0.23,Good,E,VS1,56.9,65.0,327,4.05,4.07,2.31
3,0.29,Premium,I,VS2,62.4,58.0,334,4.20,4.23,2.63
4,0.31,Good,J,SI2,63.3,58.0,335,4.34,4.35,2.75


In [ ]:
# separating features and target

In [ ]:
# Display the features (input variables)

,carat,cut,color,clarity,depth,table,x,y,z
0,0.23,Ideal,E,SI2,61.5,55.0,3.95,3.98,2.43
1,0.21,Premium,E,SI1,59.8,61.0,3.89,3.84,2.31
2,0.23,Good,E,VS1,56.9,65.0,4.05,4.07,2.31
3,0.29,Premium,I,VS2,62.4,58.0,4.20,4.23,2.63
4,0.31,Good,J,SI2,63.3,58.0,4.34,4.35,2.75
...,...,...,...,...,...,...,...,...,...
53935,0.72,Ideal,D,SI1,60.8,57.0,5.75,5.76,3.50
53936,0.72,Good,D,SI1,63.1,55.0,5.69,5.75,3.61
53937,0.70,Very Good,D,SI1,62.8,60.0,5.66,5.68,3.56
53938,0.86,Premium,H,SI2,61.0,58.0,6.15,6.12,3.74


In [ ]:
# Perform one-hot encoding on categorical features in X and drop the first category to avoid multicollinearity.
# getting dummies and dropping first


In [ ]:
# data after getting dummy variables

,carat,depth,table,x,y,z,cut_Premium,cut_Very Good,cut_Good,cut_Fair,...,color_H,color_I,color_J,clarity_VVS1,clarity_VVS2,clarity_VS1,clarity_VS2,clarity_SI1,clarity_SI2,clarity_I1
0,0.23,61.5,55.0,3.95,3.98,2.43,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0
1,0.21,59.8,61.0,3.89,3.84,2.31,1,0,0,0,...,0,0,0,0,0,0,0,1,0,0
2,0.23,56.9,65.0,4.05,4.07,2.31,0,0,1,0,...,0,0,0,0,0,1,0,0,0,0
3,0.29,62.4,58.0,4.20,4.23,2.63,1,0,0,0,...,0,1,0,0,0,0,1,0,0,0
4,0.31,63.3,58.0,4.34,4.35,2.75,0,0,1,0,...,0,0,1,0,0,0,0,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
53935,0.72,60.8,57.0,5.75,5.76,3.50,0,0,0,0,...,0,0,0,0,0,0,0,1,0,0
53936,0.72,63.1,55.0,5.69,5.75,3.61,0,0,1,0,...,0,0,0,0,0,0,0,1,0,0
53937,0.70,62.8,60.0,5.66,5.68,3.56,0,1,0,0,...,0,0,0,0,0,0,0,1,0,0
53938,0.86,61.0,58.0,6.15,6.12,3.74,1,0,0,0,...,1,0,0,0,0,0,0,0,1,0


In [ ]:
# Split the data into training and testing sets

In [ ]:
# Initialize and train the Linear Regression model

# prediction on training and testing


Linear Regression Results:
RMSE training: 1128.5697163572002
RMSE training: 1135.2116445190627


In [ ]:
# Calculate metrics


In [ ]:
# Print RMSE scores for both training and testing
# datasets to evaluate model performance.


Linear Regression Results:
RMSE training: 1128.5697163572002
RMSE test: 1135.2116445190625


In [ ]:
# Print R-squared scores for both training and testing
# datasets to evaluate model performance.


R2 score training: 0.919994314909212
R2 testing: 0.9189331350419387


In [ ]:
# Print mae scores for both training and testing
# datasets to evaluate model performance.


MAE training: 742.2563516523807
MAE test: 737.1513665933269


### Model Evaluator (e.g. RMSE, R2, MAE etc.)

- RMSE 1135 tells us that on an average our prediction will be off by 1135
- R2 score or coefficient of determination, it is a value ranging from 0-1, it tells you how much variance in the target variable is explained by the features. In our case we have r2 score of 0.91 which is extremely good.
- MAE tells us the deviation between the actual and predicted value, in our case we have a MAE of 737 which says that the deviation between actual and predicted value is 737

### Model Evaluator (e.g. RMSE, R2, MAE etc.)

In [ ]:
# Display the intercept (constant term) of the linear regression model.


array([5845.41976163])

### How much price will change with the change of X/Feature-variables? (e.g. Slope)

In [ ]:
# Display the coefficients (slopes) of each feature
# in the linear regression model as a flattened array.


array([ 1.12807843e+04, -6.50910151e+01, -2.66000210e+01, -1.00804160e+03,
       -3.52845036e+00, -3.64633703e+01,  2.62317252e+02,  1.85429483e+02,
        1.53453486e+02, -4.70152565e+00, -5.96498695e+02,  8.34883892e+02,
        6.16685289e+02,  5.55167489e+02,  3.39302365e+02, -1.64202516e+02,
       -6.44700578e+02, -1.53713594e+03,  1.54596651e+03,  1.19531483e+03,
        1.13823337e+03,  7.59927458e+02,  4.43637553e+02, -1.44563530e+02,
       -1.11853811e+03, -3.81997808e+03])